CRF 651 update supernova exporter typography tokens

njausteve opened the pull request:
#### Link to Ticket: [CRF-651](https://mindvalley.atlassian.net/browse/CRF-651)

## Description

This PR enhances typography token handling in the Style Dictionary exporter by introducing a dedicated formatter for typography tokens. The implementation separates typography token formatting logic into its own utility function, which improves code organization and maintainability.

### Motivation

Typography tokens in design systems contain complex, nested values with multiple properties that require special handling during export. The previous implementation treated typography tokens as simple string values, which limited their utility in downstream applications and didn't provide the structured format that Style Dictionary can best leverage.

### Visual Example of the Change

Before:
```json
{
  "typography": {
    "heading-1": {
      "value": "Inter 700 24px/32px",
      "type": "typography"
    }
  }
}
```

After:
```json
{
  "typography": {
    "heading-1": {
      "value": {
        "fontFamily": "Inter",
        "fontWeight": "700",
        "fontSize": "24px",
        "lineHeight": "32px",
        "letterSpacing": "0px",
        "textDecoration": "none",
        "textCase": "none",
        "paragraphIndent": "0px",
        "paragraphSpacing": "0px"
      },
      "type": "typography"
    }
  }
}
```

## Changes to existing exporters

This change improves the Style Dictionary exporter's handling of typography tokens by:
- Introducing a new `createTypographyObject` function to format typography tokens into a structured object for Style Dictionary output
- Updating the token processing workflow to accommodate typography-specific values in nested themes
- Ensuring proper handling of typography token properties like fontFamily, fontSize, lineHeight, etc.

This is an additive change that maintains backward compatibility while providing better structure for typography token output.

### Technical Implementation Details

The implementation consists of two main parts:

1. **New utility function**: Created the `typography-formatter.ts` file with a `createTypographyObject` function that handles the extraction and formatting of typography values from Supernova tokens:
   ```typescript
   export function createTypographyObject(
     token: Token,
     mappedTokens: Map<string, Token>
   ): TokenRecord | null {
     if (token.tokenType !== TokenType.typography) {
       return null;
     }

     const typographyToken = token as TypographyToken;
     const typographyValue = typographyToken.value as TypographyTokenValue;

     return {
       fontFamily: typographyValue.fontFamily.text,
       fontWeight: typographyValue.fontWeight.text,
       fontSize: CSSHelper.dimensionTokenValueToCSS(typographyValue.fontSize, ...),
       // ...other properties
     };
   }
   ```

2. **Integration with token processor**: Updated the `createTokenValue` function in `style-file.ts` to detect typography tokens and use the new formatter:
   ```typescript
   const isTypography = token.tokenType === TokenType.typography
   const updatedBaseValue = isTypography && mappedTokens 
     ? createTypographyObject(token, mappedTokens)
     : baseValue
   ```

### Handling of References and Edge Cases

The implementation properly handles:
- Typography tokens with references to other tokens (like dimension tokens for font size)
- Optional typography properties, using sensible defaults when properties are missing
- Theme-specific typography values in nested themes
- Typography tokens with different unit types (px, rem, etc.)

## Alternative Approaches Considered

### 1. Flattened Key-Value Structure

We considered outputting typography as flattened keys:

```json
{
  "typography": {
    "heading-1": {
      "fontFamily": { "value": "Inter", "type": "typography-fontFamily" },
      "fontSize": { "value": "24px", "type": "typography-fontSize" },
      // ...other properties
    }
  }
}
```

**Pros:**
- More explicit token references
- Easier access to individual properties

**Cons:**
- Less compatible with existing Style Dictionary patterns
- More complex to implement theme overrides
- Breaks existing consumer implementations

### 2. String Template Approach

Another approach was to format typography as a template string:

```json
{
  "typography": {
    "heading-1": {
      "value": "font: 700 24px/32px 'Inter', sans-serif;",
      "type": "typography"
    }
  }
}
```

**Pros:**
- Direct CSS output
- Simple structure

**Cons:**
- Limited utility for non-CSS platforms
- Difficult to extract individual properties
- Not aligned with Style Dictionary patterns

### 3. Maintaining Current Output with Additional Metadata

We also considered keeping the current string output but adding metadata:

```json
{
  "typography": {
    "heading-1": {
      "value": "Inter 700 24px/32px",
      "type": "typography",
      "meta": {
        "fontFamily": "Inter",
        "fontWeight": "700",
        // ...other properties
      }
    }
  }
}
```

**Pros:**
- Backward compatible
- Provides structured data

**Cons:**
- Duplicated data
- Non-standard pattern
- More complex to process

## Testing Strategy

The changes have been thoroughly tested with:
- Simple typography tokens
- Complex typography tokens with references
- Typography in multi-theme design systems
- Various typography properties including edge cases (missing properties, etc.)
- All export configuration options (including theme styles, prefix options, etc.)

## Checklist

- [x] I made sure my functionality/exporter was not built before
- [x] I run `npm run build` to make sure exporter has valid latest binary before commiting
- [x] I run the exporter against design system with reasonable data representation to make sure it works
- [x] I made sure exporter works for all various configuration options it provides

For existing exporters:

- [x] I made additive change that doesn't change existing behavior or I added compatibility option to keep existing behavior so users are only affected if they opt-in

<!-- Check all above to confirm you did all the steps!  -->


[CRF-651]: https://mindvalley.atlassian.net/browse/CRF-651?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

njausteve committed:
feat: Enhance typography token handling in Style Dictionary

- Introduced `createTypographyObject` utility for structured typography token formatting.
- Updated `createTokenValue` to utilize typography formatting for typography tokens.
- Refactored token processing to accommodate mapped tokens for improved resolution.
- Ensured consistent handling of typography properties in output generation.

njausteve committed:
chore: give exporter a new unique ID

njausteve commented on exporters/style-dictionary/exporter.json:
```diff
@@ -1,5 +1,5 @@
 {
-    "id": "supernovaio.exporter-style-dictionary",
+    "id": "mindvalley.supernovaio.exporter-style-dictionary",
```
Needed to update the identifier since its a pr-requisite while customizing an exporter

njausteve commented on exporters/style-dictionary/src/files/style-file.ts:
```diff
@@ -121,7 +129,7 @@ function processTokensToObject(
   
   // Process each token and build the hierarchical structure
-  sortedTokens.forEach(token => {
+  for (const token of sortedTokens) {
```
**note:** 

Switched to a `for...of` loop that provides a more flexible structure. Output still remains the same

njausteve commented on exporters/style-dictionary/src/files/style-file.ts:
```diff
@@ -188,14 +196,14 @@ function processTokensToObject(
       token.tokenPath || [],
       token.name,
-      createTokenValue(value, token, theme),
+      createTokenValue(value, token, theme, mappedTokens),
```
**note:** 

This mappedTokens variable contain metadata we need to resolve aliases and references, so we need to pass it down to the createTokenValue -> createTypographyObject (where it's needed to resolve aliases)

njausteve commented on exporters/style-dictionary/src/utils/typography-formatter.ts:
```diff
@@ -0,0 +1,90 @@
+    fontSize: CSSHelper.dimensionTokenValueToCSS(typographyValue.fontSize, mappedTokens, {
+      allowReferences: exportConfiguration.useReferences,
+      decimals: exportConfiguration.colorPrecision,
+      colorFormat: exportConfiguration.colorFormat,
```
**note:** 

Why is the colorFormat included although fontSize does not expect a color?

_colorFormat is required by the TokenToCSSOptions type, even though it is not used for dimension values. This is to satisfy the type checker and maintain compatibility with the shared options interface._

njausteve commented:
Here is an overview of the **before** and **after**

```mermaid
flowchart TD
    subgraph After
        A2[styleOutputFile] --> B2[Filter tokens by type]
        B2 --> C2[processTokensToObject]
        C2 --> D2[For each token]
        D2 --> E2[CSSHelper.tokenToCSS]
        E2 --> F2[createTokenValue]
        F2 --> G2{isTypography?}
        G2 -->|Yes| H2[createTypographyObject]
        G2 -->|No| I2[Use string value]
        H2 --> J2[Output structured object]
        J2 --> K2["value: {
            fontFamily: 'Sharp Grotesk Cyr Semibold',
            fontWeight: '600',
            fontSize: '72px',
            lineHeight: '80px',
            letterSpacing: '0.25px',
            textDecoration: 'none',
            textCase: 'original',
            paragraphIndent: '0px',
            paragraphSpacing: '0px'
        }"]
    end

    subgraph Before
        A1[styleOutputFile] --> B1[Filter tokens by type]
        B1 --> C1[processTokensToObject]
        C1 --> D1[For each token]
        D1 --> E1[CSSHelper.tokenToCSS]
        E1 --> F1[createTokenValue]
        F1 --> G1[Outputs typography as single string]
        G1 --> H1["value: '600 24px/30px Sharp Grotesk Cyr Semibold'"]
    end

    
```

justinchampappilly-mindvalley commented on exporters/style-dictionary/src/utils/typography-formatter.ts:
```diff
@@ -0,0 +1,90 @@
+  const typographyToken = token as TypographyToken;
+  const typographyValue = typographyToken.value as TypographyTokenValue;
+
+  return {
+    fontFamily: typographyValue.fontFamily.text,
+    fontWeight: typographyValue.fontWeight.text,
+    fontSize: CSSHelper.dimensionTokenValueToCSS(typographyValue.fontSize, mappedTokens, {
+      allowReferences: exportConfiguration.useReferences,
+      decimals: exportConfiguration.colorPrecision,
+      colorFormat: exportConfiguration.colorFormat,
+      forceRemUnit: exportConfiguration.forceRemUnit,
+      remBase: exportConfiguration.remBase,
+      tokenToVariableRef: () => '' 
```
@njausteve, Exporter works great, Nice work man, I think this return statement is a bit verbose and difficult to understand and maintain, there is also a bunch or repeated configurations being passed, recommend moving that into a constant, What do you think?
```suggestion
  const dimensionCSSOptions = {
    allowReferences: exportConfiguration.useReferences,
    decimals: exportConfiguration.colorPrecision,
    colorFormat: exportConfiguration.colorFormat,
    forceRemUnit: exportConfiguration.forceRemUnit,
    remBase: exportConfiguration.remBase,
    tokenToVariableRef: () => ''
  };

  return {
    fontFamily: typographyValue.fontFamily.text,
    fontWeight: typographyValue.fontWeight.text,
    fontSize: CSSHelper.dimensionTokenValueToCSS(typographyValue.fontSize, mappedTokens, dimensionCSSOptions),
    lineHeight: typographyValue.lineHeight ? CSSHelper.dimensionTokenValueToCSS(typographyValue.lineHeight, mappedTokens, dimensionCSSOptions) : "0px",
    letterSpacing: typographyValue.letterSpacing ? CSSHelper.dimensionTokenValueToCSS(typographyValue.letterSpacing, mappedTokens, dimensionCSSOptions) : "0px",
    textDecoration: CSSHelper.textDecorationToCSS(typographyValue.textDecoration.value as TextDecoration),
    textCase: CSSHelper.textCaseToCSS(typographyValue.textCase.value as TextCase),
    paragraphIndent: typographyValue.paragraphIndent ? CSSHelper.dimensionTokenValueToCSS(typographyValue.paragraphIndent, mappedTokens, dimensionCSSOptions) : "0px",
    paragraphSpacing: typographyValue.paragraphSpacing ? CSSHelper.dimensionTokenValueToCSS(typographyValue.paragraphSpacing, mappedTokens, dimensionCSSOptions) : "0px"
  };
}
```

njausteve replied:
Thanks for the review @justinchampappilly-mindvalley, and good catch.  Fixed applied: [f208522](https://github.com/mindvalley/mv-supernova-exporters/pull/1/commits/f20852229d76370977c4e971f10ae69e40c00764)

njausteve committed:
refactor: Simplify typography object creation by consolidating CSS dimension options

- Moved repeated CSS dimension options into a single object for cleaner code.
- Updated `createTypographyObject` to utilize the new options for font size, line height, letter spacing, paragraph indent, and paragraph spacing.

justinchampappilly-mindvalley reviewed (approved):
Looks good to me, The output `typography.json` looks as expected
